### DISCUSSION TITLE
于内网Linux服务器上进行穿透时，外网Windows尝试ssh即报错

### DISCUSSION THREAD

**Discussion by Qemetiel** [natfrp/wiki#395] [使用问题]
`2023/10/30 09:28:11 E Tunnel/WS2 [53/1348961/634b] 连接映射目标 [127.0.0.1:3389] 失败, 请检查本地服务是否可访问: dial tcp 127.0.0.1:3389: connectex: No connection could be made because the target machine actively refused it.`

已在服务器上按照帮助文档中内容安装 Linux服务器版natfrp service 并开通隧道（TCP，ssh(22)，#46 成都电信2），但在启动器或远程管理上开启后，在外网环境下ssh连接**时**（如`ssh testuser@frp-can.top -p 26140`，enter同时即报错）均会发生报错并断开。

另一个隧道（TCP，Windows 远程桌面(3389)，#53 成都电信4）同样在尝试进行远程连接时发生报错：

`2023/10/30 09:28:41 E Tunnel/WS2 [53/1348961/634b] 连接映射目标 [127.0.0.1:3389] 失败, 请检查本地服务是否可访问: dial tcp 127.0.0.1:3389: connectex: No connection could be made because the target machine actively refused it.`

请问可能原因是什么？非常感谢！

补充1，网络环境应该没有问题，_可以正常使用如cpolar进穿透并于外网ssh控制_；

补充2，应该不是校园网的问题，已尝试过以手机流量作为外网，甚至直接连接内网（内网是某办公室wifi），都无法ssh到服务器；

补充3，服务器Linux版本 Rocky Linux 9.2，**防火墙直接关了**，移动电脑Win10和Win11都尝试了，用到的所有frp相关版本均为最新 (至10/27/2023)；

补充4，在任意机器和网络下，节点都可以ping通。

**关于日志1的补充** 
```
2023/10/30 20:59:00 I Tunnel/WangService [46/1348961/c8b4] 限速已更新: 10 Mibit/s
  Tunnel/WangService TCP 类型隧道启动成功
  Tunnel/WangService 使用 [frp-can.top:26140] 来连接到你的隧道
  Tunnel/WangService 或使用 IP 地址连接（不推荐）: [61.139.65.143:26140]
2023/10/30 20:59:00 I Tunnel/WangService [46/1348961/c8b4] 隧道启动成功
2023/10/30 21:59:05 E Tunnel/WangService [46/1348961/c8b4] 连接映射目标 [127.0.0.1:22] 失败, 请检查本地服务是否可访问: dial tcp 127.0.0.1:22: connectex: No connection could be made because the target machine actively refused it.
```
ssh报错：
```
C:\Users\22569>ssh testuser@frp-can.top -p 26140
kex_exchange_identification: Connection closed by remote host
Connection closed by 61.139.65.143 port 26140
```

**关于日志2的补充** 远程连接应该没有问题，在同一局域网下可以正常于Windows系统进行远程控制，3389端口默认，并没有修改

再次感谢解答！

**Comment by fakeboboliu** [comment]
您好，

希望您进一步补充隧道一的日志来判断，
关于您的补充2中 `直接连接内网无法ssh到服务器` 的问题，您可以尝试在内网使用 `ssh -vvv 内网地址` 来测试，可能能看到一些有用的信息。

关于隧道二的日志，
此问题通常是因为客户端无法访问到您的服务 `127.0.0.1:3389`，即远程桌面端口修改了 或 未开启远程桌面服务。

---- 21:52 UTC+8 ----

请确认是否是可以在 *开启隧道的机器上* 使用 `ssh -v testuser@127.0.0.1 -p 22222`，或者在 *开启隧道的机器上* 使用 `127.0.0.1` 连接远程桌面？

**Reply by Qemetiel** [reply]
感谢您的回复！现将日志补充如下（原问题中也补充了）
**关于日志1的补充** 
```
2023/10/30 20:59:00 I Tunnel/WangService [46/1348961/c8b4] 限速已更新: 10 Mibit/s
  Tunnel/WangService TCP 类型隧道启动成功
  Tunnel/WangService 使用 [frp-can.top:26140] 来连接到你的隧道
  Tunnel/WangService 或使用 IP 地址连接（不推荐）: [61.139.65.143:26140]
2023/10/30 20:59:00 I Tunnel/WangService [46/1348961/c8b4] 隧道启动成功
2023/10/30 21:48:51 E Tunnel/WangService [46/1348961/c8b4] 连接映射目标 [127.0.0.1:22222] 失败, 请检查本地服务是否可访问: dial tcp 127.0.0.1:22222: connectex: No connection could be made because the target machine actively refused it.
```
ssh报错：
```
C:\Users\22569>ssh testuser@frp-can.top -p 26140
kex_exchange_identification: Connection closed by remote host
Connection closed by 61.139.65.143 port 26140
```

**关于日志2的补充** 远程连接应该没有问题，在同一局域网下可以正常于Windows系统进行远程控制，3389默认端口并没有修改。

`ssh -vvv 内网地址` 因为现在不在机器内网环境，明天会尝试，届时可能还需麻烦您。

感激不尽！

**Reply by fakeboboliu** [reply]
> 连接映射目标 [127.0.0.1:22222] 失败, 请检查本地服务是否可访问: dial tcp 127.0.0.1:22222: connectex: No connection could be made because the target machine actively refused it.
> 远程连接应该没有问题，在同一局域网下可以正常于Windows系统进行远程控制，3389默认端口并没有修改。

请确认是否是可以在 *开启隧道的机器上* 使用 `ssh -v testuser@127.0.0.1 -p 22222`，或者在 *开启隧道的机器上* 使用 `127.0.0.1` 连接远程桌面？

两个都是连接目标失败，让人感觉非常像是 IP 设置错了。

**Reply by Qemetiel** [reply]
首先抱歉之前没有把隧道指定的22222端口改回22（曾尝试调整ssh port 22改为22222，无果，又改回去了，现在确认服务器上ssh端口是22无误。）补充port 22的报错：
`2023/10/30 21:59:05 E Tunnel/WangService [46/1348961/c8b4] 连接映射目标 [127.0.0.1:22] 失败, 请检查本地服务是否可访问: dial tcp 127.0.0.1:22: connectex: No connection could be made because the target machine actively refused it.`
其他内容一致。

明天我会在 开启隧道的机器上 使用ssh测试！感谢

关于 IP 设置错了，我这里有个猜想QAQ 在 服务-远程管理 中，我的目标设备名称似乎是我当前使用的Windows设备QAQ 这是正常的吗，我是不是犯了极愚蠢的错误
(以及怎么办QAQ 麻烦您QAQQQ)

**Reply by fakeboboliu** [reply]
> 关于 IP 设置错了，我这里有个猜想QAQ 在 服务-远程管理 中，我的目标设备名称似乎是我当前使用的Windows设备QAQ 这是正常的吗

您好，这是正常的，说明您在控制的是当前使用的 Windows 设备，所以您需要把目标 IP 设置为您在使用正在控制的设备上访问相应服务时使用的 IP。

您可以尝试切换管理的设备，找到您的服务器主机名。

**Reply by Qemetiel** [reply]
QAQQQ 低级错误非常抱歉，但现在在 选择目标设备 时下方选项只有_我当前使用的设备_，并没有希望连接到的服务器主机名，这种情况应该怎么办呢（在之前的配置中缺少了哪一步吗）_(:3」∠)_

感谢您的耐心  土下座谢罪 Orz

**Reply by fakeboboliu** [reply]
> 在 选择目标设备 时下方选项只有我当前使用的设备，并没有希望连接到的服务器主机名

这个问题通常是因为：
 - 您没有打开远程控制（客户端中如果没有打开远程控制并设置 e2ee 密码，自然就不会有）
 - 客户端（启动器/服务）已经被关闭了
 - 您的几个设备主机名相同（对于同一个主机名的设备，列表中只会显示最后一个开启的）

**Reply by Qemetiel** [reply]
关于这个
1. 我有在服务器上运行./natfrp-service remote-kdf 并将得到的 e2e 密码填入natfrp-service/config.json。明天再去检查一下qwq
2. 当前使用的设备 (Windows) 上的启动器（客户端）是运行着的（且 运行模式：系统服务）；服务器上是否启动服务同样明天去检查(>_<)
3. 改了Windows主机名称，我重启一下试试（不过他俩应该不会恰好都叫做DESKTOP-VEQFV8U吧qwqqqqq）

谢谢谢谢！

**Reply by fakeboboliu** [reply]
> 1. 我有在服务器上运行./natfrp-service remote-kdf 并将得到的 e2e 密码填入natfrp-service/config.json。明天再去检查一下qwq

请记得检查一下是否设置了 `"remote_management": true,`，这是远程控制开关。

如果您希望在内网管理服务器穿透，而不是通过远程管理，也可以设置 `webui_host` 项，具体可以参考 [用户手册](https://doc.natfrp.com/launcher/manual.html#config-webui)。

> 2. 当前使用的设备 (Windows) 上的启动器（客户端）是运行着的（且 运行模式：系统服务）；服务器上是否启动服务同样明天去检查(>_<)

所以说您可以选择目标设备时下方选项有当前使用的设备，这很合理。

> 3. 改了Windows主机名称，我重启一下试试（不过他俩应该不会恰好都叫做DESKTOP-VEQFV8U吧qwqqqqq）

一般来说 Linux 服务器不会被配置为这样的 Windows 自动生成风格主机名，如果您不认为自己配置过的话，那应该不存在这个问题。

**Reply by Qemetiel** [reply]
谢谢谢谢qwq！

**Reply by Qemetiel** [reply]
汇报最新进展
在内网服务器上重新生成了.config/natfrp-service/.config.json并编辑了：
```
"token": ... 
"remote_management": true,
```
新生成的 .config.json 中不包含 remote_management_key 变量所以自己写了一个进去
**·** 一个问题是通过运行 `natfrp-service remote-kdf` 得到的 e2e 密钥结尾是 `=` （即 `......fC0uxPoIc=[root@lo......` 这样）这个 `=` 需要填入吗（感觉 = 不属于 Base64 于是没填）

之后运行
```
systemctl enable --now natfrp.service
systemctl status natfrp.service
journalctl -u natfrp.service -f
```
是有类似`远程管理成功`输出的，status里也是active，远程目标设备选择中也有此内网服务器设备项了。但在我尝试连接远程设备时网页报错为`E2E 密钥错误, 无法解密对端握手请求: Error: wrong secret key for the given ci....`

应该没有填错 e2e码... 但仍尝试修改，可能是因为修改过程中曾运行了`systemctl stop natfrp.service` 的缘故？之后设备便变得完全无法连接（应该是下线了），多次运行如 `systemctl start natfrp.service` 与 `systemctl enable --now natfrp.service` 均无果。这种情况怎么办QAQ

status 输出：
```
● natfrp.service - SakuraFrp Launcher
     Loaded: loaded (/etc/systemd/system/natfrp.service; enabled; preset: disabled)
     Active: activating (auto-restart) (Result: exit-code) since Tue 2023-10-31 01:22:54 EDT; 3s ago
    Process: 2546106 ExecStart=/home/natfrp/natfrp-service --daemon (code=exited, status=1/FAILURE)
   Main PID: 2546106 (code=exited, status=1/FAILURE)
        CPU: 39ms
```

后续！
更改了远程密码并重新得到e2e；但仍未active。
尝试了 _帮助文档-Linux 常见问题-登出后隧道断开、启动器退出_ 内容，未果。

**Reply by fakeboboliu** [reply]
> 感觉 = 不属于 Base64 于是没填
> E2E 密钥错误
> 应该没有填错 e2e码...

现在有感觉到似乎有点不对劲吗（滑稽.jpg

如果您实在对于这些配置项理解有困难，建议您考虑配置 webui 然后直接内网 webui 操作，直观方便的操作我们都放在那里了。

服务器教程是考虑到公网服务器的，希望尽可能少的暴露服务到公网，同时预期了看教程的是服务器用户，
但内网服务器可以考虑直接配置 `"webui_host": "0.0.0.0"`，然后用 `https://内网IP:4101` 管理，不要用不熟悉的东西折磨自己。

**Reply by Qemetiel** [reply]
不胜感激！终于成功了！（用远程控制

感谢您的耐心回复和指正！
